### PR TITLE
iso-tar-content: redirect private content to a signed url .

### PR DESCRIPTION
**Pull request by ldjebran**
Fixes # THEEDGE-2347

## Type of change

What is it?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Documentation update
- [ ] Tests update
- [X] Refactor

# Checklist:

- [X] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [X] I run `make pre-commit` to check fmt/vet/lint/test-no-fdo

### CODE REVIEW

File: pkg/services/files/uploader.go [hunk 1]
```diff
@@ -187,19 +187,23 @@ func (u *S3Uploader) UploadRepo(src string, account string) (string, error) {
 	return s3URL, nil
 }
 
-// UploadFile takes a Filename path as a string and then uploads that to
-// the supplied location in s3
-func (u *S3Uploader) UploadFile(fname string, uploadPath string) (string, error) {
+// uploadFile takes a Filename path as a string and then uploads that to
+// uploadFileWithACL  upload a file from local file system to remote s3 bucket location using the acl supplied.
+func (u *S3Uploader) uploadFileWithACL(fname string, uploadPath string, acl string) (string, error) {
```
Comment (omaciel) [inline pkg/services/files/uploader.go:174]: Are there tests for this module/function?

File: pkg/routes/imagesets.go [hunk 2]
```diff
@@ -36,6 +36,10 @@ func MakeImageSetsRouter(sub chi.Router) {
 	})
 }
 
+func getStorageInstallerIsoURL(installerID uint) string {
```
Comment (omaciel) [inline pkg/routes/imagesets.go:39]: How about a simple test?

**Review by omaciel — CHANGES_REQUESTED** [review]
Basically asking for more tests, as we have code changes that imho should have tests in place, as well as new code that will benefit from having new tests while everything is still fresh in your mind :)

File: pkg/routes/imagesets.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/routes/imagesets.go:39]: done

File: pkg/routes/imagesets.go [hunk 3]
```diff
@@ -204,7 +208,8 @@ func ListAllImageSets(w http.ResponseWriter, r *http.Request) {
 					result = db.DB.First(&i.Installer, &i.InstallerID)
 				}
 				if i.Installer.ImageBuildISOURL != "" {
-					imgSet.ImageBuildISOURL = &i.Installer.ImageBuildISOURL
+					imageSetInstallerIsoURL := getStorageInstallerIsoURL(i.Installer.ID)
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/routes/imagesets.go]: done

File: pkg/routes/imagesets.go [hunk 4]
```diff
@@ -279,6 +284,13 @@ func GetImageSetsByID(w http.ResponseWriter, r *http.Request) {
 	details.ImageSetData = *imageSet
 	details.Images = Imgs
 
+	// update image installer iso URL for all images with the internal application storage end-point
+	for _, imageDetail := range details.Images {
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/routes/imagesets.go:295]: done

File: pkg/routes/imagesets.go (hunk 2 above)
Comment (ldjebran) [reply pkg/routes/imagesets.go:39]: done

File: pkg/services/files/downloader.go [hunk 5]
```diff
@@ -36,10 +49,55 @@ func (d *HTTPDownloader) DownloadToPath(sourceURL string, destinationPath string
...
+			Bucket: aws.String(cfg.BucketName),
+			Key:    aws.String(url.Path),
+		})
+
+	if err != nil {
+		d.log.WithField("error", err.Error()).Errorf(`Error downloading file : "%s"`, url.Path)
+		return err
+	}
+
+	d.log.Infof(`sourceURL: "%s" successfully downloaded to destinationPath: "%s"`, sourceURL, destinationPath)
```
Comment (AaronH88) [inline pkg/services/files/downloader.go:101]: I think this should be `d.log.Debug` else we will have lots of info messages when we redirect a `/repo`

File: pkg/services/files/downloader.go [hunk 6]
```diff
@@ -36,10 +49,55 @@ func (d *HTTPDownloader) DownloadToPath(sourceURL string, destinationPath string
...
 	return err
 }
+
+// S3Downloader aws s3 files Downloader, download resources at aws s3 bucket via sdk
+type S3Downloader struct {
+	log *log.Entry
+}
+
+// DownloadToPath download function that puts the source_url  at s3 bucket into the destination_path on the local filesystem
+func (d *S3Downloader) DownloadToPath(sourceURL string, destinationPath string) error {
```
Comment (AaronH88) [inline pkg/services/files/downloader.go:66]: Is this function thread safe? 

Can I run this in 100 `go` workers? My main concern is the `os.Create()` command? Is that thread safe?

File: pkg/services/files/s3.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/redhatinsights/edge-api/logger"
+)
+
+// GetNewS3Session return a new aws s3 session
+func GetNewS3Session() *session.Session {
+	cfg := config.Get()
+	var sess *session.Session
+	if cfg.Debug {
```
Comment (AaronH88) [inline pkg/services/files/s3.go:16]: If we deploy the app in stage with DEBUG=TRUE will this session be able to access S3?

if we need a new param for local development then I suggest adding a new `Dev` config field

File: pkg/services/files/uploader.go [hunk 8]
```diff
@@ -214,3 +200,8 @@ func (u *S3Uploader) UploadFile(fname string, uploadPath string) (string, error)
 	s3URL := fmt.Sprintf("https://%s.s3.%s.amazonaws.com/%s", u.Bucket, region, uploadPath)
 	return s3URL, nil
 }
+
+// UploadFile takes a Filename path as a string and then uploads that to the supplied location in s3
+func (u *S3Uploader) UploadFile(fname string, uploadPath string) (string, error) {
```
Comment (AaronH88) [inline pkg/services/files/uploader.go:205]: This func is a great idea! Love it

File: pkg/services/repobuilder.go [hunk 9]
```diff
@@ -251,12 +250,23 @@ func (rb *RepoBuilder) DownloadVersionRepo(c *models.Commit, dest string) (strin
...
+	if external {
+		rb.log.WithFields(log.Fields{"filepath": tarFileName, "imageBuildTarURL": c.ImageBuildTarURL}).Debug("Grabbing tar file")
+		_, err = grab.Get(tarFileName, c.ImageBuildTarURL)
+
+		if err != nil {
+			rb.log.WithField("error", err.Error()).Error("Error grabbing tar file")
+			return "", err
+		}
+	} else {
+		rb.log.WithFields(log.Fields{"filepath": tarFileName, "imageBuildTarURL": c.ImageBuildTarURL}).Debug("Downloading tar file")
```
Comment (AaronH88) [inline pkg/services/repobuilder.go:263]: I think this can be a debug message

File: pkg/services/files/downloader.go (hunk 6 above)
Comment (ldjebran) [reply pkg/services/files/downloader.go:66]: The os.Create should be an atomic os operation it success or fail, the s3 download manager is thread safe  when writing to file.

File: pkg/services/files/s3.go (hunk 7 above)
Comment (ldjebran) [reply pkg/services/files/s3.go:16]: Yes, as we are not changing the behaviour of the s3 session , but only moving the code to an independent location , as the same code got repeated again and again.

File: pkg/services/files/downloader.go (hunk 5 above)
Comment (AaronH88) [reply pkg/services/files/downloader.go:101]: Happy with this answer and to keep it the way it currently is

File: pkg/services/files/s3.go [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+	"github.com/redhatinsights/edge-api/config"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/redhatinsights/edge-api/logger"
+)
+
+// GetNewS3Session return a new aws s3 session
+func GetNewS3Session() *session.Session {
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/files/s3.go:13]: done

File: pkg/services/files/downloader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/files/downloader.go:66]: This is the same as for for GetSigned url , will create new issue to investigate on testing s3 functions

File: pkg/services/files/uploader.go (hunk 1 above)
Comment (ldjebran) [reply pkg/services/files/uploader.go:174]: This is the same as for for GetSigned url , will create new issue to investigate on testing s3 functions

Comment (ldjebran) [reply pkg/services/files/uploader.go:174]: but also here we have a private function

File: pkg/services/files.go [hunk 11]
```diff
@@ -110,7 +100,13 @@ func (s *LocalFilesService) GetFile(path string) (io.ReadCloser, error) {
 	return f, nil
 }
 
-// GetFile retuns the file given a path
+// GetSignedURL return a signed URL
+func (s *LocalFilesService) GetSignedURL(path string) (string, error) {
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/files.go:104]: done

File: pkg/services/files.go [hunk 12]
```diff
@@ -110,7 +100,13 @@ func (s *LocalFilesService) GetFile(path string) (io.ReadCloser, error) {
...
+	// locale File service does not support signed url, and return the same original url
+	return path, nil
+}
+
+// GetFile returns the file given a path
 func (s *S3FilesService) GetFile(path string) (io.ReadCloser, error) {
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/files.go:110]: will open an issue about testing app s3 functions

File: pkg/services/files.go [hunk 13]
```diff
@@ -132,3 +128,19 @@ func (s *S3FilesService) GetFile(path string) (io.ReadCloser, error) {
 	}
 	return o.Body, nil
 }
+
+// defaultURLSignatureExpiry The default URL signature expiry time in minutes
+const defaultURLSignatureExpiry = 120
+
+// GetSignedURL return and aws s3 bucket signed url
+func (s *S3FilesService) GetSignedURL(path string) (string, error) {
```
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/files.go:136]: will open an issue about testing app s3 functions

File: pkg/services/files/downloader.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ldjebran) [reply pkg/services/files/downloader.go:66]: will open an issue about testing app s3 functions

**Comment by ldjebran** [comment]
@AaronH88 @omaciel openned an issue THEEDGE-2388 about investigating on unit testing s3 app functions

**Review by AaronH88 — APPROVED** [review]
/lgtm

**Review by AaronH88 — APPROVED** [review]
/lgtm

**Pull request merged**